### PR TITLE
samples: net: add put and post methods to coap

### PR DESCRIPTION
**Pull request by tarunkum**
In sample application of COAP put and post
    methods are implemented in coap-server.c
    and coap-client.c

Signed-off-by: Tarun Kumar Agarwal tarun.kumar.agarwal@intel.com

### CODE REVIEW

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Please add {} even for one liners. The original code did not had these but if you make changes then it is worth adding those braces.

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Indentation is wrong either above or below, so please fix accordingly.

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Please do not put length when defining the array of this kind, just let the compiler calculate the length for you.

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Is there a reason why the length is specified here, for example token1 is only 7 chars long. Could you omit the length?

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Put the curly brace to same line as switch statement.

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Return indentation is wrong

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Fix indentation. Also there should be space after ",". You could check these issues before submitting the patch, there is scripts/checkpatch.pl script just for this purpose.

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Fix indentation

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: { in wrong line

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: No c++ comments

File: samples/net/coap_client/src/coap-client.c
Comment (jukkar) [inline samples/net/coap_client/src/coap-client.c]: Just remove dead code if it is not used.

**Review by jukkar — CHANGES_REQUESTED** [review]
Please use scripts/checkpatch.pl to check your patch before submitting again, your patch has lot of style issues.

Also your Signed-off line is wrong, it should be your real email address. See example syntax in git log.

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: paload -> payload

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: I believe Zephyr is not using CamelCase naming convention.
Alter "LoopCount" to loopcount or loop_count

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: for (...) {
   ...
}

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: same here

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: use type "enum coap_method" here.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: You can call coap_next_token to get an unique token w/o declaring My_COAP_token1/2

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: If we use type "enum coap_method" in the function interface, then we could use "COAP_METHOD__XXX" here as case id.

**Review by rbtchc — CHANGES_REQUESTED** [review]

**Comment by tarunkum** [comment]
Hello Jukka/Robert,

Thanks a lot for review. I have done the changes as per the given comments. I have uploaded the new changes in the same branch.

Regards,
Tarun

**Comment by tarunkum** [comment]
Hello @jukkar / @rbtchc ,

I have removed all the warnings.

**Comment by rbtchc** [comment]
@tarunkum you still need to fix sanity test failures. Following the linked posted by Shippable at the bottom of the PR. Also, there are still many code styling issues, you might want to fix those as well.

**Comment by tarunkum** [comment]
@rbtchc , Looking into will try to solve all warnings as soon as possible.

**Comment by tarunkum** [comment]
Hello @jukkar / @rbtchc ,

All the issues has resolved now. Please let me know if any changes are required.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: this is no longer a callback function registered and called when an interface is up. please update the interface to remove the unnecessary parameters and change the name to avoid misleading

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: type "u8_t" -> "enum coap_method"
and perhaps we should call it "method" instead of "code"?

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: is it necessary to have different payload contents for different requests?

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: restore this line

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: remove this line

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: remove this line

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: line 237 ~ 256 can be simplified as
```

if (code == COAP_METHOD_DELETE) {
        printk("Delete method is not supported");
        return;
}

r = coap_packet_init(&request, pkt, 1, COAP_TYPE_CON,
                                  8, coap_next_token(), code,
                                  coap_next_id());
```

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: if we consolidate two payloads variable into one then we can replace the following as
`
if (code == COAP_METHOD_PUT || code == CODE_METHOD_POST) {
		r = coap_packet_append_payload_marker(&request);
		if (r) {
			net_pkt_unref(pkt);
			printk("Unable to append payload marker\n");
		}

		r = coap_packet_append_payload(&request, (u8_t *)put_payload,
				                                       strlen(put_payload));
		if (r < 0) {
			net_pkt_unref(pkt);
			printk("Not able to append payload\n");
		}
}
`

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: restore line 275 and line 277

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: remove this line. only one blank line is required between function declarations.

**Review by rbtchc — CHANGES_REQUESTED** [review]

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: I want to use different payload message for both methods. So that I have separated out.

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: I have to keep break this function into two then only I can test other methods at a same time. Otherwise it is not possible.

	event_iface_init();
	printk("\n ************ COAP Client  test 1 *************\n");
	event_iface_up(NULL, NET_EVENT_IF_UP, iface, COAP_METHOD_GET);

So that I have implemented. if you want to give new name to event_iface_init(); function let me know.

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: Added the case for COAP_METHOD_DELETE method.

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: done

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: Further implemented with Semaphore now.

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: Removed the test variable in msg_dump function

**Comment by tarunkum** [comment]
Hello @rbtchc ,

All the issues has resolved now. Please let me know if any further changes are required.

**Comment by tarunkum** [comment]
@rbtchc I am waiting for this patch to be merged. Please let me know if any other change is required in code.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: Suggest to change the function name to something like: send_coap_request(...)
And remove "struct net_if *iface" for you are not using it.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: I suppose this should be named in lower and do not append comments after the code.

**Review by rbtchc — CHANGES_REQUESTED** [review]
I suppose you should ask @jukkar or @rveerama1 to review the patch as well for I'm not the code owner.

**Comment by rveerama1** [comment]
I am fine if you address @rbtchc comments.

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: Modified the code for this comment.

**Comment by tarunkum** [comment]
@rveerama1 I have addressed all the review comments. Please look into once.

**Comment by tarunkum** [comment]
@rbtchc I have addressed all the review comments. Please look into once.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: Shouldn't we abort the sending request when payload append fail?

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [inline samples/net/coap_client/src/coap-client.c]: Shouldn't we abort the sending request when payload append fail?

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: paylod -> payload

File: samples/net/coap_server/src/coap-server.c
Comment (rbtchc) [inline samples/net/coap_server/src/coap-server.c]: paylod -> payload

File: samples/net/coap_server/prj.conf
Comment (rbtchc) [inline samples/net/coap_server/prj.conf]: is this necessary?

**Review by rbtchc — CHANGES_REQUESTED** [review]

File: samples/net/coap_client/src/coap-client.c
Comment (tarunkum) [reply samples/net/coap_client/src/coap-client.c]: @rbtchc Thanks I have addressed in new code changes. Please review it once.

**Comment by tarunkum** [comment]
@rbtchc Please review the code. I have addressed all the review comments.

File: samples/net/coap_client/src/coap-client.c
Comment (rbtchc) [reply samples/net/coap_client/src/coap-client.c]: @tarunkum where?

**Comment by tarunkum** [comment]
@rbtchc I have merged the requested changes. Please review once.

**Review by rbtchc — COMMENTED** [review]
LGTM

**Comment by tarunkum** [comment]
@rveerama1 

Robert has reviewed & validated.

Please review one & let me know if any changes are required.

**Review by rveerama1 — APPROVED** [review]

**Comment by tarunkum** [comment]
@kumarvikash1 solved the merge conflict. Rebased the code.

**Pull request merged**
